Use upstream ANALYZE code.

hlinnaka opened the pull request:
This replaces most of the Greenplum code used to ANALYZE a relation,
with the upstream implementation.

This should make merging with upstream easier. This should also speed
up ANALYZE considerably, as collect the sample of rows and calculate the
stats for each column in-memory, insteead of issuing queries for each one.

This changes the VacAttrStats to also match upstream. That's part of the
public API, for writing custom analyze functions. Hence the update to
the PostGIS submodule - I had to revert the GPDB-specific changes to
analyze-functions in PostGIS .

hlinnaka commented:
For context, see:

*  thread "Differences in statistics collected between PostgreSQL and Greenplum" on gpdb-dev: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/zGVGkUMFytQ/9-DLbNDnCgAJ.
* pull request #241
* issue #321
* issue #279

d commented on .gitmodules:
```diff
@@ -16,7 +16,7 @@
 [submodule "gpAux/extensions/postgis-2.0.3/source"]
 	path = gpAux/extensions/postgis-2.0.3/source
-	url = ssh://git@stash.eng.pivotal.io:2222/gpdb/postgis.git
+	url = ssh://git@stash.eng.pivotal.io:2222/~hlinnakangas/postgis.git
```
are you sure of including this line of change?

hlinnaka replied:
On 29/01/16 03:55, Jesse Zhang wrote:
>> @@ -16,7 +16,7 @@
>>   	branch = master
>>   [submodule "gpAux/extensions/postgis-2.0.3/source"]
>>   	path = gpAux/extensions/postgis-2.0.3/source
>> -	url = ssh://git@stash.eng.pivotal.io:2222/gpdb/postgis.git
>> +	url = ssh://git@stash.eng.pivotal.io:2222/~hlinnakangas/postgis.git
>
> are you sure of including this line of change?

Ah, right, that needs some more discussion. (No, clearly we shouldn't 
make the submodule to point at my private repository ;-) )

The reason that this patch, as submitted, needs changes to PostGIS is 
that it changes the VacAttrStats struct. That struct is part of the 
public interface for writing a type analyze function, so we need to fix 
the compute_geography_stats() and compute_geometry_stats() functions in 
PostGIS accordingly.

I'd really like to make the proposed changes to VacAttrStats, because it 
actually changes the struct *back* the way it is in the upstream, and 
the changes to PostGIS *remove* GPDB-specific changes there, reverting 
those functions the way they are in upstream PostGIS. However, it may be 
better to do that as a separate patch, for the sake of easier review. So 
I'll change this patch to not change VacAttrStats in an incompatible 
way, and I'll make a separate pull request for that.

Hmm, actually, VacAttrStats doesn't seem to be used for anything in 
current git master. There are no calls to the type-specific "typanalyze" 
function at all. So apparently the changes to the VacAttrStat struct, 
and the PostGIS typanalyze functions, are completely superfluous anyway.

Thanks!

- Heikki

hlinnaka commented:
Ok, I opened pull request #328 for the VacAttrStats changes. I intend to commit that first, and this bigger patch after that.

hlinnaka commented:
@vraghavan78, @hsyuan, I would greatly appreciate review of this. I would love to get this done, as this significantly shortens the time it takes to run the regression tests, and I believe @changleicn is also eagerly waiting for this, to cherry-pick it into HAWQ.

hsyuan commented:
looks good to me.
`analyzeutils.h` & `analyzeutils.c` can be removed, since they are not used by other files any more.

hlinnaka commented:
There's one regression test failure with this, in "subselect_gp" test case, because of a changed plan, with ORCA. I'm not entirely sure why ORCA chooses a different plan for this, because AFAICS the cost estimates for the tables are the same in both plans. But the new plan is totally reasonable, perhaps even better than the old one, so this seems harmless. Detail below:

```

regression=# \d subselect_t1 
 Table "public.subselect_t1"
 Column |  Type   | Modifiers 
--------+---------+-----------
 x      | integer | 
Distributed by: (x)

regression=# \d subselect_t2
 Table "public.subselect_t2"
 Column |  Type   | Modifiers 
--------+---------+-----------
 y      | integer | 
Distributed by: (y)

regression=# select * from subselect_t1;
 x 
---
 1
 2
(2 rows)

regression=# select * from subselect_t2;
 y 
---
 2
 2
 1
(3 rows)

-- BEFORE PR #321 ---

regression=# select * from pg_stats where tablename like 'subselect_%';
 schemaname |  tablename   | attname | null_frac | avg_width | n_distinct | most_common_vals |  most_common_freqs  | histogram_bounds | correlation 
------------+--------------+---------+-----------+-----------+------------+------------------+---------------------+------------------+-------------
 public     | subselect_t1 | x       |         0 |         4 |         -1 | {1,2}            | {0.5,0.5}           |                  |            
 public     | subselect_t2 | y       |         0 |         4 |  -0.666667 | {2,1}            | {0.666667,0.333333} |                  |            
(2 rows)

regression=# explain  select  count(*) from subselect_t1 where x in (select y from subselect_t2);
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..862.00 rows=1 width=8)
   ->  Gather Motion 2:1  (slice1; segments: 2)  (cost=0.00..862.00 rows=1 width=8)
         ->  Aggregate  (cost=0.00..862.00 rows=1 width=8)
               ->  Hash Join  (cost=0.00..862.00 rows=1 width=1)
                     Hash Cond: subselect_t1.x = subselect_t2.y
                     ->  Table Scan on subselect_t1  (cost=0.00..431.00 rows=1 width=4)
                     ->  Hash  (cost=431.00..431.00 rows=1 width=4)
                           ->  GroupAggregate  (cost=0.00..431.00 rows=1 width=4)
                                 Group By: subselect_t2.y
                                 ->  Sort  (cost=0.00..431.00 rows=2 width=4)
                                       Sort Key: subselect_t2.y
                                       ->  Table Scan on subselect_t2  (cost=0.00..431.00 rows=2 width=4)
 Settings:  optimizer=on
 Optimizer status: PQO version 1.621
(14 rows)


-- AFTER PR #321 ---

regression=# select * from pg_stats where tablename like 'subselect_%';
 schemaname |  tablename   | attname | null_frac | avg_width | n_distinct | most_common_vals | most_common_freqs | histogram_bounds | correlation 
------------+--------------+---------+-----------+-----------+------------+------------------+-------------------+------------------+-------------
 public     | subselect_t1 | x       |         0 |         4 |         -1 |                  |                   | {1,2}            |           1
 public     | subselect_t2 | y       |         0 |         4 |  -0.666667 | {2}              | {0.666667}        |                  |           1
(2 rows)

regression=# explain  select  count(*) from subselect_t1 where x in (select y from subselect_t2);
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..862.00 rows=1 width=8)
   ->  Gather Motion 2:1  (slice1; segments: 2)  (cost=0.00..862.00 rows=1 width=8)
         ->  Aggregate  (cost=0.00..862.00 rows=1 width=8)
               ->  Hash EXISTS Join  (cost=0.00..862.00 rows=1 width=1)
                     Hash Cond: subselect_t1.x = subselect_t2.y
                     ->  Table Scan on subselect_t1  (cost=0.00..431.00 rows=1 width=4)
                     ->  Hash  (cost=431.00..431.00 rows=2 width=4)
                           ->  Table Scan on subselect_t2  (cost=0.00..431.00 rows=2 width=4)
 Settings:  optimizer=on
 Optimizer status: PQO version 1.621
(10 rows)

```

vraghavan78 commented:
The plan change is okay. But the change is the stats is not easy to digest. For column y we have lost the information that there is another data point "1". I think the table level sampling was introduced by a former greenplum employee to reduce the error in statistics derivation. I am afraid, we may be re-introducing that error.

hlinnaka commented:
On 08/02/16 22:44, Venkatesh Raghavan wrote:
> The plan change is okay it could be a tie in the memo. But the change
> is the stats is concerning. For column y we have lost the information
> that there is another data point "1". I think the table level
> sampling was introduced by a former greenplum employee to reduce the
> error in statistics derivation. I am afraid, we may be re-introducing
> that error.

You do realize that there are exactly three rows in the table? You can't 
derive very meaningful stats from that.

There's not much need to remember the there was one "1" value in the 
table, as one row is the minimum estimate anyway.

PS. What is "table level sampling"?

- Heikki

vraghavan78 commented:
>> You do realize that there are exactly three rows in the table? You can't derive very meaningful stats >> from that.

I agree. I just wanted to know why the stats changed. 

Table sampling is uniform-random sampling over the table rather than block-level sampling.

hlinnaka commented:
On 09/02/16 00:14, Venkatesh Raghavan wrote:
> Table sampling is uniform-random sampling over the table rather than block-level sampling.

Ok. I haven't changed that in this patch, i.e. it still uses 
uniform-random sampling over the table. I actually think that 
block-level sampling would be better, as scanning the whole table is 
really slow, but the PostgreSQL code that does that would be laborious 
to adapt to append-only tables and distributed tables. So I kept GPDB 
method of collecting the sample with a "SELECT ... WHERE random() < 
threshold" query. But the code that derives the statistics from that 
sample is replaced with the upstream code.

- Heikki

hlinnaka commented:
I pushed PR #328, which contained the PostGIS changes. That addresses @d's comment, this patch no longer contains the postgis change (as has now been committed as part of #328). Here's a new version:

* Rebased on top of the current master. 
* Removed analyzeutil.c and analyzeutil.h (Thanks, @hsyuan!)
* Added a note to the commit message explaining that we still use the full-table-scanning sampling query to collect the sample

hsyuan commented:
LGTM.

hlinnaka commented:
Pushed, thanks for the review!
